Enable SparkSqlHook to use supplied connections

danielenricocahall opened the pull request:
In case of existing issue, reference it using one of the following:

closes: https://github.com/apache/airflow/issues/8713

Minor update to the logic in `SparkSqlHook` to follow same behavior of `SparkSubmitHook` by determining master and connection parameters through connection ID.

potiuk commented on airflow/providers/apache/spark/hooks/spark_sql.py:
```diff
@@ -67,27 +67,50 @@ def __init__(
         executor_memory: Optional[str] = None,
         keytab: Optional[str] = None,
         principal: Optional[str] = None,
-        master: str = 'yarn',
```
This is not a backwards-compatible change. Can you please make it works also if someone calls it in the old way ? We have some mechanism (just look for "deprecated") that we mark certain parameters as deprecated, but the rule of thumb we have is that we deprecate first and only after next major release we remove the deprecated features. It seems possible to add (and test) the old way of using the hook here.

Also the docstrings were not updated - but they should be - including the deprecation information and explanation what has been replaced.

danielenricocahall replied:
@potiuk reverted to make backwards compatible and add deprecation explanation in docstrings!

potiuk commented on airflow/providers/apache/spark/hooks/spark_sql.py:
```diff
@@ -67,27 +67,50 @@ def __init__(
+            else:
+                conn_data['master'] = conn.host
+
+            # Determine optional yarn queue from the extra field
```
Can you plese describe this new behaviour in the documentation (including deprecated way, marking it as such? ) we have  a newly restructured documentation and it is now very easy to find out relevant part of the documentation (just rebase to latest master and see the `docs/apache-airflow-providers-apache-spark'.

danielenricocahall replied:
@potiuk added some pieces in documentation - how does it look?

danielenricocahall committed:
refactor such that spark sql hook and operator utilizes conn_id

fix tests

refactor such that spark sql hook and operator utilizes conn_id

linting/static analysis

linting/static analysis again

refactor such that spark sql hook and operator utilizes conn_id

linting/static analysis

refactor such that spark sql hook and operator utilizes conn_id

linting

refactor such that spark sql hook and operator utilizes conn_id

parameterize tests

refactor such that spark sql hook and operator utilizes conn_id

michalslowikowski00 commented on tests/providers/apache/spark/hooks/test_spark_sql.py:
```diff
@@ -213,3 +209,29 @@ def test_spark_process_runcmd_and_fail(self, mock_popen):
         )
+
+    def test_resolve_connection_yarn_default_connection(self):
+        hook = SparkSqlHook(conn_id='spark_default', sql='SELECT 1')
```
This can be put into setup() test method, after that you will be have access through `self` to the hook and you will avoid code duplication. Also connection = hook._resolve_connection() can be put in setUp().
```
    def setUp(self):
            self.hook = SparkSqlHook(conn_id='spark_yarn_cluster', sql='SELECT 1')
            self.connection = self.hook._resolve_connection()
```

michalslowikowski00 replied:
I would go even further and use https://github.com/wolever/parameterized. 
This is basically the same test case with different data `expected_spark_connection` that can be parametrized thorough @parameterized(). 
Just saying. :)

danielenricocahall replied:
Great idea! Yeah I am familiar with the concept as pytest has `parametrize` which I use. I will make those changes and resubmit sometime today.

danielenricocahall replied:
@michalslowikowski00 done!

michalslowikowski00 replied:
Cool :D

ashb commented on airflow/providers/apache/spark/hooks/spark_sql.py:
```diff
@@ -44,15 +44,19 @@ class SparkSqlHook(BaseHook):
     :param verbose: Whether to pass the verbose flag to spark-sql
     :type verbose: bool
-    :param yarn_queue: The YARN queue to submit to (Default: "default")
+    :param yarn_queue: (Deprecated) The YARN queue to submit to (Default: "default")
```
You haven't just deprecated this -- you've made it do nothing at all.

Please make this argument still be respected if passed

ashb commented on airflow/providers/apache/spark/operators/spark_sql.py:
```diff
@@ -56,6 +58,8 @@ class SparkSqlOperator(BaseOperator):
     :param verbose: Whether to pass the verbose flag to spark-sql
     :type verbose: bool
     :param yarn_queue: The YARN queue to submit to (Default: "default")
+        This parameter has been deprecated. Master and connection parameters (such as YARN queue)
```
I don't think we want to deprecate this --  it is a valid option to override the queue for just a single operator for instance.

ashb reviewed (changes requested):
You haven't deprecated, you've made the parameters do nothing at all.

We probably don't want to deprecate entirely either -- there is plenty of precedent for overloading some values on a per-operator basis in other operators.
